libr/bin/dwarf.c

```diff
@@ -1346,293 +1647,495 @@ static const ut8 *r_bin_dwarf_parse_attr_value(const ut8 *obuf, int obuf_len,
-			for (j = 0; j < value->encoding.block.length; j++) {
-				value->encoding.block.data[j] = READ (buf, ut8);
+		if (value->block.data) {
+			for (j = 0; j < value->block.length; j++) {
+				value->block.data[j] = READ (buf, ut8);
 			}
 		}
 		break;
-	case DW_FORM_block1:
-		value->encoding.block.length = READ (buf, ut8);
-		value->encoding.block.data = calloc (sizeof (ut8), value->encoding.block.length + 1);
-		if (value->encoding.block.data) {
-			for (j = 0; j < value->encoding.block.length; j++) {
-				value->encoding.block.data[j] = READ (buf, ut8);
+	case DW_FORM_block2:
+		value->block.length = READ16 (buf);
+		if (value->block.length > 0) {
+			value->block.data = calloc (sizeof (ut8), value->block.length);
+			if (!value->block.data) {
+				return NULL;
+			}
+			for (j = 0; j < value->block.length; j++) {
+				value->block.data[j] = READ (buf, ut8);
 			}
 		}
 		break;
-	case DW_FORM_flag:
-		value->encoding.flag = READ (buf, ut8);
+	case DW_FORM_block4:
+		value->block.length = READ32 (buf);
+		if (value->block.length > 0) {
+			ut8 *data = calloc (sizeof (ut8), value->block.length);
+			if (!data) {
+				return NULL;
+			}
+			for (j = 0; j < value->block.length; j++) {
+				data[j] = READ (buf, ut8);
+			}
+			value->block.data = data;
+		}
 		break;
-	case DW_FORM_sdata:
-		buf = r_leb128 (buf, buf_end - buf, &value->encoding.sdata);
+	case DW_FORM_block: // variable length ULEB128
+		buf = r_uleb128 (buf, buf_end - buf, &value->block.length);
+		if (!buf || buf >= buf_end) {
+			return NULL;
+		}
+		value->block.data = calloc (sizeof (ut8), value->block.length);
+		if (value->block.data) {
+			for (j = 0; j < value->block.length; j++) {
+				value->block.data[j] = READ (buf, ut8);
+			}
+		}
 		break;
-	case DW_FORM_strp:
-		value->encoding.str_struct.offset = READ32 (buf);
-		if (debug_str && value->encoding.str_struct.offset < debug_str_len) {
-			value->encoding.str_struct.string = strdup (
+	case DW_FORM_flag:
+		value->flag = READ (buf, ut8);
+		break;
+	case DW_FORM_strp: // offset in .debug_str
+		// this offset can be 64bit, based on dwarf format
+		if (hdr->is_64bit) {
+			value->string.offset = READ64 (buf);
+		} else {
+			value->string.offset = READ32 (buf);
+		}
+		if (debug_str && value->string.offset < debug_str_len) {
+			// TODO does it make sense to duplicate the
+			value->string.content = strdup (
 				(const char *)(debug_str +
-					value->encoding.str_struct.offset));
+					value->string.offset));
 		} else {
-			value->encoding.str_struct.string = NULL;
+			value->string.content = NULL;
 		}
 		break;
-	case DW_FORM_udata:
-		{
-			ut64 ndata = 0;
-			const ut8 *data = (const ut8*)&ndata;
-			buf = r_uleb128 (buf, R_MIN (sizeof (data), (size_t)(buf_end - buf)), &ndata);
-			memcpy (&value->encoding.data, data, sizeof (value->encoding.data));
-			value->encoding.str_struct.string = NULL;
+	case DW_FORM_ref_addr: // offset in .debug_info
+		// This is 4 or 8 bytes depending where it refers to
```

Comment:
Please fix the comment indentation.

Reply:
yeah those fixes belong into the previous PR that this PR builds on, I'll merge the changes from there after